Deploy staging instance on PR

ryanmohta opened the pull request:
This PR updates the GitHub Actions workflow to deploy a staging instance when a PR is created or updated. This staging instance will be hosted at `staging-{ pr number }.peterportal.org`.

To accomplish this, I had to dynamically create the `DOMAIN` environment variable to include the staging prefix if the source of the workflow run was a pull request. This environment variable would then be used in the `api/serverless.yaml` file to indicate which domain to deploy the project to. All credit for this process goes to @Voark who provided guidance on how to accomplish this!

I haven’t gotten to test this out yet, as the only way to do so is to actually make the PR. I’ll keep making changes if it doesn’t happen to be correct, and only request reviews once I believe it’s ready to be merged.

### To-do
- [x] Add comment indicating to the user what URL the app was deployed to
- [x] Clean up the stack when the PR is closed
- [x] Add concurrency when multiple commits to a PR are made in quick succession

ryanmohta commented:
Closing this PR to test the “remove the stack when PR is closed/merged” functionality

ryanmohta commented:
Ready for prime-time (I hope)! Marking this as ready for review and sending out reviews

Voark commented on api/serverless.yaml:
```diff
@@ -21,7 +20,8 @@ provider:
   region: us-east-1
-  stage: prod
+  stage: ${env:STACK_NAME}
+  stackName: peterportal-${env:STACK_NAME}
```
Shouldn't this be `peterportal-core-${env:STACK_NAME}`? Not entirely sure how serverless works but this is based off the current stack name I see for your prod stack.

Voark replied:
Otherwise when you deploy to prod you'll make a new stack instead of updating the current one.

ryanmohta replied:
Oh that’s right, I’ll go ahead and update that!

ryanmohta replied:
Just fixed this so I’ll resolve this conversation

Voark reviewed (commented):
Besides my comment, LGTM!

Also for some inexplicable reason the first few times I opened the staging site on my current browser (Arc) I got an error in the console, but it was working fine on other browsers (FF, Chrome). I just went back to try and screenshot the error but now it works fine on Arc too. Not a big deal just wanted to mention it...

ryanmohta commented:
Ah you bring up a good point about concurrency, I’ll look more into that and see if there’s a way to have the second action cancel the first one if both of them run simultaneously

Voark commented:
> Ah you bring up a good point about concurrency, I’ll look more into that and see if there’s a way to have the second action cancel the first one if both of them run simultaneously.
> 
> Edit: I looked at the [GitHub docs](https://docs.github.com/en/actions/using-jobs/using-concurrency) and there’s a couple ways to handle concurrency: waiting for the first job to finish before starting the second one, or canceling the first job in-progress and starting the second one immediately. Which do you think works better for our deployments?

For AntAlmanac we cancel the current build and start a new run. However this introduces a new issue where if the previous run already started deploying, it can't be stopped on a whim so there is a [step](https://github.com/icssc/AntAlmanac/blob/3b236ff59d506795d264b9f18152486ff2889a19/.github/workflows/stage_pr.yaml#L60) that waits for the stack resources to finish updating (if they are). I added this because if you did back to back to back commits you wouldn't build up an action queue, but it does add some complexity.

IMO, I would do cancel in progress. If deploying with Serverless automatically waits for the stack to stabilize it should work well, if not feel free to copy paste our [stabilize action](https://github.com/icssc/AntAlmanac/blob/main/.github/actions/stabilize_stack/action.yaml), should be plug and play.

ryanmohta commented:
Just added concurrency; I didn’t see any way to stop the serverless deployment while it was in progress, so I added a part to the GitHub Action to keep running the deployment step if the job was canceled. This way, if commit B is pushed before commit A gets to the deployment step, commit A gets canceled, but if commit A is currently deploying, it finishes the deployment while the GitHub Action for commit B starts. Even though this will lead to slightly overlapping GitHub Action runs, based on how long our build process takes it won’t lead to overlapping _deployments_.

Voark commented:
> Just added concurrency; I didn’t see any way to stop the serverless deployment while it was in progress, so I added a part to the GitHub Action to keep running the deployment step if the job was canceled. This way, if commit B is pushed before commit A gets to the deployment step, commit A gets canceled, but if commit A is currently deploying, it finishes the deployment while the GitHub Action for commit B starts. Even though this will lead to slightly overlapping GitHub Action runs, based on how long our build process takes it won’t lead to overlapping _deployments_.

Yeah there is no way to stop the an aws stack from being deployed, which is why in our action I wait for the deployment to finish before trying a new one. Also forgot to mention you probably also want to add concurrency to the destroy action otherwise you will have to manually delete stacks if it runs while your stack is currently in deployment.

ryanmohta commented:
Closing PR once again for testing

ryanmohta commented:
Made some updates to how concurrency is handled: to make it easier to isolate the deployment step from the rest of the workflow (and in good coding practice) I’ve split the workflow into a _build_ job and a _deploy_ job, each with their own concurrency group. If two builds jobs overlap, the second job will cancel the first, while if two deploy jobs overlap, the second job will wait for the first one to finish. This is so that we don’t accidentally cancel in the middle of a deployment, which could corrupt the site.

I made use of the GitHub Actions cache to pass the built files from the build job to the deploy job, and ensured uniqueness between different caches by assigning keys of `${{ github.workflow }} #${{ github.run_number }} attempt ${{ github.run_attempt }}` (for example, `Build and deploy #9 attempt 1`). This way the deployment can use the same cache as the build for the same workflow run, but deployments or builds from two different workflow runs (or even two different attempts of the same run) will get their own cache.

The "Clean up PR” workflow uses the same concurrency group as the deploy job, since they both make changes to AWS. Deployments and clean-ups will run one after the other instead of one canceling the other, just to be safe.

ryanmohta commented:
Re-requesting a review from @Voark since I fixed all the changes you suggested!

Voark reviewed (approved):
Just a small comment on the action, otherwise LGTM! 

I do think the infrastructure should still be looked into to make deploying and building easier/more efficient (obviously out of scope of this PR), I'll open an issue for it sometime.

ryanmohta commented on .github/workflows/clean-up-pr.yml:
```diff
@@ -0,0 +1,39 @@
+  clean-up-pr:
+    runs-on: ubuntu-latest
+    env:
+      STAGING_PREFIX: ${{ github.event_name == 'pull_request' && format('staging-{0}.', github.event.pull_request.number) || '' }}
```
good catch, must be me copy-pasting from the build and deploy action and forgetting to change this here. I’ll go fix this now!

ethanwong16 commented:
Closes #255

ethanwong16 reviewed (approved):
LGTM 🚀 🚀 🚀 🚀 🚀 !
Awesome job @ryanmohta  and thanks @Voark for the help!
